### PR TITLE
Handle new TypeScript version

### PR DESCRIPTION
**Pull request by mia-pi-git**
This is a PR for the reason that the new TS changes how try-catch handles types. Previously, it would assume in the  `catch (x)` that `x` was any. Now, it assumes it's unknown. Likewise, it can only be cast to `unknown` or `any`.
I'd like to hear people's opinions on if we should:
a) Keep it as `unknown`, and handle case-by-case.
b) Cast to any, and leave all cases as it was before.
My current approach is of casting to `any`, and handling as we did prior. Please share your thoughts if you have an opinion.

### CODE REVIEW

File: server/chat-plugins/hangman.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ try {
 	if (save) {
 		FS(HANGMAN_FILE).writeUpdate(() => JSON.stringify(hangmanData));
 	}
-} catch (e) {}
+} catch (e: any) {}
```
Comment (AnnikaCodes) [inline server/chat-plugins/hangman.ts:40]: Why does _this_ need a cast?

File: server/chat-plugins/helptickets.ts [hunk 2]
```diff
@@ -793,7 +793,7 @@ export async function getBattleLog(battle: string): Promise<BattleInfo | null> {
 				url: `https://${Config.routes.replays}/${battle}`,
 			};
 		}
-	} catch (e) {}
+	} catch (e: any) {}
```
Comment (AnnikaCodes) [inline server/chat-plugins/helptickets.ts:796]: Or this?

File: server/chat-plugins/hangman.ts (hunk 1 above)
Comment (mia-pi-git) [reply server/chat-plugins/hangman.ts:40]: The general response to this is that I did a mass find and replace. Might be nice to keep it as a general convention?

File: server/chat-plugins/helptickets.ts (hunk 2 above)
Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:796]: https://github.com/smogon/pokemon-showdown/pull/8449#discussion_r698140489

File: server/chat-plugins/hangman.ts (hunk 1 above)
Comment (mia-pi-git) [reply server/chat-plugins/hangman.ts]: I pushed a commit to remove the types from all patterns of `catch (e) {}`.

File: lib/crashlogger.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ export function crashlogger(
 		if (!transport) {
 			try {
 				require.resolve('nodemailer');
-			} catch (e) {
+			} catch (e: any) {
```
Comment (Slayer95) [inline lib/crashlogger.ts:54]: Might as well use [optional catch binding](https://mariusschulz.com/blog/optional-catch-binding-in-typescript).

**Comment by Zarel** [comment]
I'm surprised this requires a PR. In theory there should be a config setting to make it always glocal.

**Comment by Zarel** [comment]
Yeah, just disable `useUnknownInCatchVariables` in tsconfig.

**Comment by Zarel** [comment]
Although I guess it's an open question if we'd prefer the setting or the explicit `any` type.

File: lib/crashlogger.ts (hunk 3 above)
Comment (Zarel) [reply lib/crashlogger.ts:54]:
```suggestion
			} catch {
```

File: lib/crashlogger.ts [hunk 4]
```diff
@@ -65,7 +65,7 @@ export function crashlogger(
 		} else {
 			try {
 				transport = require('nodemailer').createTransport(Config.crashguardemail.options);
-			} catch (e) {
+			} catch (e: any) {
```
Comment (Zarel) [inline lib/crashlogger.ts:68]:
```suggestion
			} catch {
```

File: lib/sql.ts [hunk 5]
```diff
@@ -51,7 +51,7 @@ type ErrorHandler = (error: Error, data: DatabaseQuery) => void;
 function getModule() {
 	try {
 		return require('better-sqlite3') as typeof sqlite.default;
-	} catch (e) {
+	} catch (e: any) {
```
Comment (Zarel) [inline lib/sql.ts:54]:
```suggestion
	} catch {
```

File: server/chat-commands/admin.ts [hunk 6]
```diff
@@ -114,7 +114,7 @@ async function updateserver(context: Chat.CommandContext, codePath: string) {
 		}
 
 		return true;
-	} catch (e) {
+	} catch (e: any) {
```
Comment (Zarel) [inline server/chat-commands/admin.ts:117]:
```suggestion
	} catch {
```

File: server/chat-commands/info.ts [hunk 7]
```diff
@@ -2475,7 +2475,7 @@ export const commands: Chat.ChatCommands = {
 		if (!/^(https?:\/\/)?(www\.)?(youtube\.com|youtu\.be)(\/|$)/i.test(link)) {
 			try {
 				dimensions = await Chat.fitImage(link);
-			} catch (e) {
+			} catch (e: any) {
```
Comment (Zarel) [inline server/chat-commands/info.ts:2478]:
```suggestion
			} catch {
```

File: server/chat-commands/info.ts [hunk 8]
```diff
@@ -2589,7 +2589,7 @@ export const commands: Chat.ChatCommands = {
 				const [width, height, resized] = await Chat.fitImage(link);
 				buf = Utils.html`<img src="${link}" width="${width}" height="${height}" />`;
 				if (resized) buf += Utils.html`<br /><a href="${link}" target="_blank">full-size image</a>`;
-			} catch (err) {
+			} catch (err: any) {
```
Comment (Zarel) [inline server/chat-commands/info.ts:2592]:
```suggestion
			} catch {
```

File: server/chat-plugins/battlesearch.ts [hunk 9]
```diff
@@ -47,7 +47,7 @@ export async function runBattleSearch(userids: ID[], month: string, tierid: ID,
 		let output;
 		try {
 			output = await ProcessManager.exec(['rg', '-i', regexString, '--no-line-number', '-P', '-tjson', ...files]);
-		} catch (error) {
+		} catch (error: any) {
```
Comment (Zarel) [inline server/chat-plugins/battlesearch.ts:50]:
```suggestion
		} catch {
```

File: server/chat-plugins/chatlog.ts [hunk 10]
```diff
@@ -66,7 +66,7 @@ export class LogReaderRoom {
 		try {
 			const listing = await FS(`logs/chat/${this.roomid}`).readdir();
 			return listing.filter(file => /^[0-9][0-9][0-9][0-9]-[0-9][0-9]$/.test(file));
-		} catch (err) {
+		} catch (err: any) {
```
Comment (Zarel) [inline server/chat-plugins/chatlog.ts:69]:
```suggestion
		} catch {
```

File: server/chat-plugins/chatlog.ts [hunk 11]
```diff
@@ -75,7 +75,7 @@ export class LogReaderRoom {
 		try {
 			const listing = await FS(`logs/chat/${this.roomid}/${month}`).readdir();
 			return listing.filter(file => file.endsWith(".txt")).map(file => file.slice(0, -4));
-		} catch (err) {
+		} catch (err: any) {
```
Comment (Zarel) [inline server/chat-plugins/chatlog.ts:78]:
```suggestion
		} catch {
```

File: server/chat-plugins/chatlog.ts [hunk 12]
```diff
@@ -212,7 +212,7 @@ export const LogReader = new class {
 				const days = (await FS(`logs/${month}/${tier}/`).readdir()).filter(this.isDay).sort();
 				firstDay = days[0];
 				break;
-			} catch (err) {}
+			} catch (e) {}
```
Comment (Zarel) [inline server/chat-plugins/chatlog.ts:215]:
```suggestion
			} catch {}
```

File: server/chat-plugins/chatlog.ts [hunk 13]
```diff
@@ -225,7 +225,7 @@ export const LogReader = new class {
 				const days = (await FS(`logs/${month}/${tier}/`).readdir()).filter(this.isDay).sort();
 				lastDay = days[days.length - 1];
 				break;
-			} catch (err) {}
+			} catch (e) {}
```
Comment (Zarel) [inline server/chat-plugins/chatlog.ts:228]:
```suggestion
			} catch {}
```

File: server/chat-plugins/chatlog.ts [hunk 14]
```diff
@@ -242,7 +242,7 @@ export const LogReader = new class {
 				Utils.sortBy(battles, getBattleNum);
 
 				return [getBattleNum(battles[0]), getBattleNum(battles[battles.length - 1])];
-			} catch (err) {
+			} catch (err: any) {
```
Comment (Zarel) [inline server/chat-plugins/chatlog.ts:245]:
```suggestion
			} catch {
```

File: server/chat-plugins/daily-spotlight.ts [hunk 15]
```diff
@@ -50,7 +50,7 @@ export async function renderSpotlight(roomid: RoomID, key: string, index: number
 				const [width, height] = await Chat.fitImage(image, 150, 300);
 				imgHTML = `<td><img src="${image}" width="${width}" height="${height}" style="vertical-align:middle;"></td>`;
 				spotlights[roomid][key][index].image = [image, width, height];
-			} catch (err) {}
+			} catch (e) {}
```
Comment (Zarel) [inline server/chat-plugins/daily-spotlight.ts:53]:
```suggestion
			} catch {}
```

File: server/chat-plugins/daily-spotlight.ts [hunk 16]
```diff
@@ -179,7 +179,7 @@ export const commands: Chat.ChatCommands = {
 			img = img.trim();
 			try {
 				[width, height] = await Chat.fitImage(img);
-			} catch (e) {
+			} catch (e: any) {
```
Comment (Zarel) [inline server/chat-plugins/daily-spotlight.ts:182]:
```suggestion
			} catch {
```

File: server/chat-plugins/helptickets.ts [hunk 17]
```diff
@@ -766,7 +766,7 @@ export async function getOpponent(link: string, submitter: ID): Promise<string |
 			const body = await replayUrl.get();
 			const data = JSON.parse(body);
 			return data.p1id === submitter ? data.p1id : data.p2id;
-		} catch (e) {
+		} catch (e: any) {
```
Comment (Zarel) [inline server/chat-plugins/helptickets.ts:769]:
```suggestion
		} catch {
```

**Comment by mia-pi-git** [comment]
> Although I guess it's an open question if we'd prefer the setting or the explicit `any` type.

Yeah, that's why I made this a PR (plus Wob asked me to.) I didn't know about the setting, but I think it couldn't hurt to make it open to ask. (Setting seems better though.)

**Comment by Zarel** [comment]
Fine by me.

I would recommend doing the optional `catch` binding (my suggestions, but I only did the first half). `crashlogger` should also be updated to take `unknown`, so we won't need it for anything that calls out to `crashlogger`.

**Comment by Zarel** [comment]
Looks like eslint or typescript-eslint don't support optional catch binding. We'll probably need to upgrade those.

**Comment by Zarel** [comment]
I tried to update to `@types/node@16` but apparently they do some weird `globalThis` stuff which I couldn't figure out in like an hour of googling, so back to `@types/node@14` we go.

**Comment by Zarel** [comment]
Thanks!

**Pull request merged**